modules/nlu/src/backend/application/index.ts

```diff
@@ -0,0 +1,129 @@
+    private _engine: NLU.Engine,
+    private _botFactory: IBotFactory,
+    private _botService: IBotService,
+    queueTrainingOnBotMount: boolean = true
+  ) {
+    this._queueTrainingOnBotMount = queueTrainingOnBotMount
+  }
+
+  public async initialize() {
+    await this._trainingQueue.initialize()
+  }
+
+  public get trainRepository(): ITrainingRepository {
+    return this._trainingQueue.repository
+  }
+
+  public teardown = async () => {
+    for (const botId of this._botService.getIds()) {
+      await this.unmountBot(botId)
+    }
+    return this._trainingQueue.teardown()
+  }
+
+  public getHealth() {
+    return this._engine.getHealth()
+  }
+
+  public async getTraining(botId: string, language: string): Promise<TrainingState> {
+    return this._trainingQueue.getTraining({ botId, language })
+  }
+
+  async getAllTrainings(): Promise<TrainingSession[]> {
+    return this._trainingQueue.getAllTrainings()
+  }
+
+  async pauseTrainings() {
+    return this._trainingQueue.pause()
+  }
+
+  async resumeTrainings(): Promise<void> {
+    await this._trainingQueue.resume()
+  }
+
+  public hasBot = (botId: string) => {
+    return !!this._botService.getBot(botId)
+  }
+
+  public getBot(botId: string): Predictor {
+    const bot = this._botService.getBot(botId)
+    if (!bot) {
+      throw new BotNotMountedError(botId)
+    }
+    return bot
+  }
+
+  public mountBot = async (botConfig: BotConfig) => {
+    const { id: botId, languages } = botConfig
+    const { bot, defService, modelRepo } = await this._botFactory.makeBot(botConfig)
+    this._botService.setBot(botId, bot)
+
+    const makeDirtyModelHandler = (cb: (trainId: TrainingId) => Promise<void>) => async (language: string) => {
+      const latestModelId = await defService.getLatestModelId(language)
+      if (await modelRepo.hasModel(latestModelId)) {
+        await bot.load(latestModelId)
+        return
+      }
+      return cb({ botId, language })
+    }
+
+    const loadOrSetTrainingNeeded = makeDirtyModelHandler((trainId: TrainingId) =>
+      this._trainingQueue.needsTraining(trainId)
+    )
+    defService.listenForDirtyModels(loadOrSetTrainingNeeded)
+
+    const trainingEnabled = !yn(process.env.BP_NLU_DISABLE_TRAINING)
+    const trainingHandler =
+      this._queueTrainingOnBotMount && trainingEnabled
+        ? (trainId: TrainingId) => this._trainingQueue.queueTraining(trainId)
+        : (trainId: TrainingId) => this._trainingQueue.needsTraining(trainId)
+
+    const loadModelOrQueue = makeDirtyModelHandler(trainingHandler)
+    for (const language of languages) {
+      await loadModelOrQueue(language)
+    }
+    await bot.mount()
+  }
+
+  public unmountBot = async (botId: string) => {
+    const bot = this._botService.getBot(botId)
+    if (!bot) {
+      throw new BotNotMountedError(botId)
+    }
+    await bot.unmount()
+    await this._trainingQueue.cancelTrainings(botId) // TODO: fully remove training sessions
```

Comment:
why this comment ? this leave zombie sessions hagning ?  just like when we kill a server ? 

is there a trainging session janitor to cleanup zombie trainings ? or it's fixed by design

Reply:
For now it leaves zombie session hanging until it's detected as zombie by `queueTraining` in `training-queue` by checking if it's been more than 5 minutes since the last update. We plan on adding a janitor later.